### PR TITLE
Unify the logic of fetching product by SKU

### LINKED ISSUE #11919: EAN-13 check digit not recognized when scanning barcode in Order creation
**Describe the bug**
A clear and concise description of what the bug is. Please be as descriptive as possible; issues lacking detail, or for any other reason than to report a bug, may be closed without action.
EAN-13 check digit not recognized when scanning barcode in Order creation. It's recognized correctly in Scan to update inventory though.

Discussion and context: p1720098058277949/1720098051.555859-slack-C6H8C3G23

**To Reproduce**
1. Generate EAN-13 barcode
2. Verify the last value is not recognised in Order creation
3. Verify the last value is correctly recognised in Scan to update inventory (Products screen)

**Expected behavior**
1. The last digit of the code should be recognised.
2. Scan to update inventory and Order creation should provide the same results when scanning.

### PR DESCRIPTION
**Pull request by samiuelson**
Closes: #11919 

### Description

This PR intends to unify the logic responsible for fetching the product by SKU. 

Context: Users reported a problem that it's impossible to add a product by scanning its barcode in the order creation flow, however, the product was detected correctly in the Products screen. The hypothesis was that the EAN-13 checksum was not handled correctly in the `OrderCreateEditViewModel`, however, it turned out that the product was not found because it was fetched from the local storage only.

There are also the following reasons for that change:
* Keeping the business logic extracted from view models in a use case – making it easier to read, and debug the code.
* Avoid discrepancies between how we fetch products by SKU in different places in the app (Scan to update product inventory – Products screen, and Order creation flow). 
* Avoid code duplication
* Extract/reduce code in the `OrderCreateEditViewModel` class that contains too much code

To achieve the above, this PR switches to using the existing `FetchProductBySKU` class in the `OrderCreateEditViewModel`. The difference between the existing logic in the VM and the `FetchProductBySKU` use case is that the latter on attempts to look for the product on the remote in case it's not found in the local cache.

### Steps to reproduce

1. Create a new order
2. Scan a barcode of a product that is not present in the local cache (DB)
3. Verify the app fails to add the product

### Testing information

It's important to do thorough tests of the order creation flow — adding a product by scanning the barcode, including edge-cases, e.g. wrong barcode value (not matching any SKU), barcode matching existing SKU present on remote only, etc.

### The tests that have been performed

I tested that this PR fixes the above scenario. Device: Pixel 8, Android 14

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-android/milestone/235">20.4</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/OrderCreateEditViewModel.kt [hunk 1]
```diff
@@ -904,6 +902,7 @@ class OrderCreateEditViewModel @Inject constructor(
                         barcodeFormat = status.format
                     )
                 )
+                viewState = viewState.copy(isUpdatingOrderDraft = false)
```
Comment (AnirudhBhat) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/OrderCreateEditViewModel.kt]: ❓ I don't think we need this. Please refer my comment below.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/OrderCreateEditViewModel.kt [hunk 2]
```diff
@@ -924,12 +923,20 @@ class OrderCreateEditViewModel @Inject constructor(
             }
         }.orEmpty()
         viewModelScope.launch {
-            productRepository.searchProductList(
-                searchQuery = barcodeOptions.sku,
-                skuSearchOptions = WCProductStore.SkuSearchOptions.ExactSearch,
-            )?.let { products ->
-                handleFetchProductBySKUSuccess(products, selectedItems, source, barcodeOptions)
-            } ?: run {
+            val result = fetchProductBySKU(barcodeOptions.sku, barcodeOptions.barcodeFormat)
```
Comment (AnirudhBhat) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/OrderCreateEditViewModel.kt:925]: ⚠️  Since this is launched in a coroutine, i think we need to display progress bar (isUpdatingOrderDraft = true) so that UI won't look like it is frozen. What do you think?

Also, we could write unit tests to test these business logic.

**Review by AnirudhBhat — COMMENTED** [review]
Thanks for the improvement @samiuelson. The fix works fine. I've left couple of comments about showing progress bar while we fetch product from remote and adding unit tests. Let me know what you think!

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/OrderCreateEditViewModel.kt (hunk 2 above)
Comment (samiuelson) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/OrderCreateEditViewModel.kt:925]: Done – 5e74c632ef6e966e1d7a9559e09ab28f999cfa06

Comment (samiuelson) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/OrderCreateEditViewModel.kt:925]: > we could write unit tests to test these business logic

I reviewed existing unit tests and think the logic of `fetchProductBySKU` is covered pretty well. If you can see any scenario missing let me know though, @AnirudhBhat!

**Comment by samiuelson** [comment]
Thanks for the review, @AnirudhBhat! I've addressed your comments. Let me know what you think.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/OrderCreateEditViewModel.kt (hunk 2 above)
Comment (AnirudhBhat) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/OrderCreateEditViewModel.kt:925]: I meant adding tests for the newly made changes of showing and hiding progress bar while fetching SKU.

**Review by AnirudhBhat — APPROVED** [review]
LGTM! I've just left one minor comment about adding unit test for the newly added logic of displaying and hiding progress bar while fetching product from SKU.
